TypeError: Found non-callable @@iterator on verify conditions (take 2!)

I'm still seeing this error on the latest, previously reported as #207.

`TypeError: Found non-callable @@iterator at verifyConditions`

It when running semantic-release in a private project, and when running via `npx`.

It happens in https://github.com/FauxFaux/testcase-semantic-release , although you will need to have write access to github to get to the point, making it hard to demo in docker.

```
% npm run semantic-release 

> revise@1.0.0 semantic-release /var/tmp/faux191028.revise
> semantic-release

[1:49:39 PM] [semantic-release] › ℹ  Running semantic-release version 15.13.28
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/github"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/github"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/github"
[1:49:40 PM] [semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/github"
[1:49:40 PM] [semantic-release] › ⚠  This run was not triggered in a known CI environment, running in dry-run mode.
[1:49:40 PM] [semantic-release] › ⚠  Run automated release from branch master in dry-run mode
[1:49:43 PM] [semantic-release] › ✔  Allowed to push to the Git repository
[1:49:43 PM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/npm"
[1:49:43 PM] [semantic-release] [@semantic-release/npm] › ℹ  Verify authentication for registry https://registry.npmjs.org/
[1:49:43 PM] [semantic-release] › ✖  Failed step "verifyConditions" of plugin "@semantic-release/npm"
[1:49:43 PM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/github"
[1:49:43 PM] [semantic-release] [@semantic-release/github] › ℹ  Verify GitHub authentication
[1:49:43 PM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/github"
[1:49:43 PM] [semantic-release] › ✖  An error occurred while running semantic-release: { TypeError: Found non-callable @@iterator
    at verifyConditions (/var/tmp/faux191028.revise/node_modules/@semantic-release/npm/index.js:38:12) pluginName: '@semantic-release/npm' }
{ AggregateError: 
    TypeError: Found non-callable @@iterator
        at verifyConditions (/var/tmp/faux191028.revise/node_modules/@semantic-release/npm/index.js:38:12)
    at /var/tmp/faux191028.revise/node_modules/semantic-release/lib/plugins/pipeline.js:54:11 name: 'AggregateError' }npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! revise@1.0.0 semantic-release: `semantic-release`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the revise@1.0.0 semantic-release script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/faux/.cache/npm/_logs/2019-10-28T13_49_43_571Z-debug.log
```

```
% npm ls @semantic-release/npm
revise@1.0.0 /var/tmp/faux191028.revise
└─┬ semantic-release@15.13.28
  └── @semantic-release/npm@5.3.1 
```

The error it is attempting to spread,

https://github.com/semantic-release/npm/blob/7db019f9563297670562807b44f18195bc05f013/index.js#L37-L39

is:
```
{ [Error: ENOENT: no such file or directory, stat '/var/tmp/faux191028.revise/.npmrc']
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/var/tmp/faux191028.revise/.npmrc' }
```

I do indeed not have a `.npmrc` in the current directory.

--- a/lib/set-npmrc-auth.js
+++ b/lib/set-npmrc-auth.js
@@ -1,18 +1,10 @@
 const path = require('path');
 const rc = require('rc');
-const {outputFile, copy, readFile} = require('fs-extra');
+const {outputFile, readFile} = require('fs-extra');
 const getAuthToken = require('registry-auth-token');
 const nerfDart = require('nerf-dart');
 const AggregateError = require('aggregate-error');
 const getError = require('./get-error');
-
-const readFileIfExists = async path => {
-  try {
-    return await readFile(path);
-  } catch (_) {
-    return '';
-  }
-};
 
 module.exports = async (
   npmrc,
@@ -20,17 +12,23 @@ module.exports = async (
   {cwd, env: {NPM_TOKEN, NPM_CONFIG_USERCONFIG, NPM_USERNAME, NPM_PASSWORD, NPM_EMAIL}, logger}
 ) => {
   logger.log('Verify authentication for registry %s', registry);
-  const config = NPM_CONFIG_USERCONFIG || path.resolve(cwd, '.npmrc');
-  if (getAuthToken(registry, {npmrc: rc('npm', {registry: 'https://registry.npmjs.org/'}, {config})})) {
-    await copy(config, npmrc);
+  const {configs, ...rcConfig} = rc(
+    'npm',
+    {registry: 'https://registry.npmjs.org/'},
+    {config: NPM_CONFIG_USERCONFIG || path.resolve(cwd, '.npmrc')}
+  );
+  const currentConfig = configs ? (await Promise.all(configs.map(config => readFile(config)))).join('\n') : '';
+
+  if (getAuthToken(registry, {npmrc: rcConfig})) {
+    await outputFile(npmrc, currentConfig);
     return;
   }
 
   if (NPM_USERNAME && NPM_PASSWORD && NPM_EMAIL) {
-    await outputFile(npmrc, `${await readFileIfExists(config)}\n_auth = \${LEGACY_TOKEN}\nemail = \${NPM_EMAIL}`);
+    await outputFile(npmrc, `${currentConfig}\n_auth = \${LEGACY_TOKEN}\nemail = \${NPM_EMAIL}`);
     logger.log(`Wrote NPM_USERNAME, NPM_PASSWORD and NPM_EMAIL to ${npmrc}`);
   } else if (NPM_TOKEN) {
-    await outputFile(npmrc, `${await readFileIfExists(config)}\n${nerfDart(registry)}:_authToken = \${NPM_TOKEN}`);
+    await outputFile(npmrc, `${currentConfig}\n${nerfDart(registry)}:_authToken = \${NPM_TOKEN}`);
     logger.log(`Wrote NPM_TOKEN to ${npmrc}`);
   } else {
     throw new AggregateError([getError('ENONPMTOKEN', {registry})]);

--- a/test/set-npmrc-auth.test.js
+++ b/test/set-npmrc-auth.test.js
@@ -3,72 +3,129 @@ import test from 'ava';
 import {readFile, appendFile} from 'fs-extra';
 import {stub} from 'sinon';
 import tempy from 'tempy';
-import setNpmrcAuth from '../lib/set-npmrc-auth';
+import clearModule from 'clear-module';
+
+const {HOME} = process.env;
+const cwd = process.cwd();
 
 test.beforeEach(t => {
   // Stub the logger
   t.context.log = stub();
   t.context.logger = {log: t.context.log};
+
+  clearModule('rc');
+  clearModule('../lib/set-npmrc-auth');
 });
 
-test('Set auth with "NPM_TOKEN"', async t => {
+test.afterEach.always(() => {
+  process.env.HOME = HOME;
+  process.chdir(cwd);
+});
+
+test.serial('Set auth with "NPM_TOKEN"', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
   const env = {NPM_TOKEN: 'npm_token'};
 
-  await setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
 
   t.regex((await readFile(npmrc)).toString(), /\/\/custom.registry.com\/:_authToken = \$\{NPM_TOKEN\}/);
   t.deepEqual(t.context.log.args[1], [`Wrote NPM_TOKEN to ${npmrc}`]);
 });
 
-test('Set auth with "NPM_USERNAME", "NPM_PASSWORD" and "NPM_EMAIL"', async t => {
+test.serial('Set auth with "NPM_USERNAME", "NPM_PASSWORD" and "NPM_EMAIL"', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
   const env = {NPM_USERNAME: 'npm_username', NPM_PASSWORD: 'npm_pasword', NPM_EMAIL: 'npm_email'};
 
-  await setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
 
   t.is((await readFile(npmrc)).toString(), `\n_auth = \${LEGACY_TOKEN}\nemail = \${NPM_EMAIL}`);
   t.deepEqual(t.context.log.args[1], [`Wrote NPM_USERNAME, NPM_PASSWORD and NPM_EMAIL to ${npmrc}`]);
 });
 
-test('Copy ".npmrc" if auth is already configured', async t => {
+test.serial('Preserve home ".npmrc"', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
+  const npmrc = tempy.file({name: '.npmrc'});
+  const env = {NPM_TOKEN: 'npm_token'};
+
+  await appendFile(path.resolve(process.env.HOME, '.npmrc'), 'home_config = test');
+
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
+
+  t.is((await readFile(npmrc)).toString(), `home_config = test\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`);
+  t.deepEqual(t.context.log.args[1], [`Wrote NPM_TOKEN to ${npmrc}`]);
+});
+
+test.serial('Preserve home and local ".npmrc"', async t => {
+  process.env.HOME = tempy.directory();
+  const cwd = tempy.directory();
+  process.chdir(cwd);
+  const npmrc = tempy.file({name: '.npmrc'});
+  const env = {NPM_TOKEN: 'npm_token'};
+
+  await appendFile(path.resolve(cwd, '.npmrc'), 'cwd_config = test');
+  await appendFile(path.resolve(process.env.HOME, '.npmrc'), 'home_config = test');
+
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger});
+
+  t.is(
+    (await readFile(npmrc)).toString(),
+    `home_config = test\ncwd_config = test\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`
+  );
+  t.deepEqual(t.context.log.args[1], [`Wrote NPM_TOKEN to ${npmrc}`]);
+});
+
+test.serial('Preserve all ".npmrc" if auth is already configured', async t => {
+  process.env.HOME = tempy.directory();
+  const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
 
   await appendFile(path.resolve(cwd, '.npmrc'), `//custom.registry.com/:_authToken = \${NPM_TOKEN}`);
+  await appendFile(path.resolve(process.env.HOME, '.npmrc'), 'home_config = test');
 
-  await setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger});
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger});
 
-  t.is((await readFile(npmrc)).toString(), `//custom.registry.com/:_authToken = \${NPM_TOKEN}`);
+  t.is((await readFile(npmrc)).toString(), `home_config = test\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`);
   t.is(t.context.log.callCount, 1);
 });
 
-test('Copy ".npmrc" if auth is already configured for a scoped package', async t => {
+test.serial('Preserve ".npmrc" if auth is already configured for a scoped package', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
 
   await appendFile(
     path.resolve(cwd, '.npmrc'),
     `@scope:registry=http://custom.registry.com\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`
   );
+  await appendFile(path.resolve(process.env.HOME, '.npmrc'), 'home_config = test');
 
-  await setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger});
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger});
 
   t.is(
     (await readFile(npmrc)).toString(),
-    `@scope:registry=http://custom.registry.com\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`
+    `home_config = test\n@scope:registry=http://custom.registry.com\n//custom.registry.com/:_authToken = \${NPM_TOKEN}`
   );
   t.is(t.context.log.callCount, 1);
 });
 
-test('Throw error if "NPM_TOKEN" is missing', async t => {
+test.serial('Throw error if "NPM_TOKEN" is missing', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
 
   const [error] = await t.throwsAsync(
-    setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger})
+    require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env: {}, logger: t.context.logger})
   );
 
   t.is(error.name, 'SemanticReleaseError');
@@ -76,13 +133,15 @@ test('Throw error if "NPM_TOKEN" is missing', async t => {
   t.is(error.code, 'ENONPMTOKEN');
 });
 
-test('Emulate npm config resolution if "NPM_CONFIG_USERCONFIG" is set', async t => {
+test.serial('Emulate npm config resolution if "NPM_CONFIG_USERCONFIG" is set', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
 
   await appendFile(path.resolve(cwd, '.custom-npmrc'), `//custom.registry.com/:_authToken = \${NPM_TOKEN}`);
 
-  await setNpmrcAuth(npmrc, 'http://custom.registry.com', {
+  await require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {
     cwd,
     env: {NPM_CONFIG_USERCONFIG: path.resolve(cwd, '.custom-npmrc')},
     logger: t.context.logger,
@@ -92,13 +151,15 @@ test('Emulate npm config resolution if "NPM_CONFIG_USERCONFIG" is set', async t 
   t.is(t.context.log.callCount, 1);
 });
 
-test('Throw error if "NPM_USERNAME" is missing', async t => {
+test.serial('Throw error if "NPM_USERNAME" is missing', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
   const env = {NPM_PASSWORD: 'npm_pasword', NPM_EMAIL: 'npm_email'};
 
   const [error] = await t.throwsAsync(
-    setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
+    require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
   );
 
   t.is(error.name, 'SemanticReleaseError');
@@ -106,13 +167,15 @@ test('Throw error if "NPM_USERNAME" is missing', async t => {
   t.is(error.code, 'ENONPMTOKEN');
 });
 
-test('Throw error if "NPM_PASSWORD" is missing', async t => {
+test.serial('Throw error if "NPM_PASSWORD" is missing', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
   const env = {NPM_USERNAME: 'npm_username', NPM_EMAIL: 'npm_email'};
 
   const [error] = await t.throwsAsync(
-    setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
+    require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
   );
 
   t.is(error.name, 'SemanticReleaseError');
@@ -120,13 +183,15 @@ test('Throw error if "NPM_PASSWORD" is missing', async t => {
   t.is(error.code, 'ENONPMTOKEN');
 });
 
-test('Throw error if "NPM_EMAIL" is missing', async t => {
+test.serial('Throw error if "NPM_EMAIL" is missing', async t => {
+  process.env.HOME = tempy.directory();
   const cwd = tempy.directory();
+  process.chdir(cwd);
   const npmrc = tempy.file({name: '.npmrc'});
   const env = {NPM_USERNAME: 'npm_username', NPM_PASSWORD: 'npm_password'};
 
   const [error] = await t.throwsAsync(
-    setNpmrcAuth(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
+    require('../lib/set-npmrc-auth')(npmrc, 'http://custom.registry.com', {cwd, env, logger: t.context.logger})
   );
 
   t.is(error.name, 'SemanticReleaseError');

